### PR TITLE
Refactor loadbalancer features

### PR DESCRIPTION
**Pull request by spencerhance**
Features are now defined with structs instead of nested maps

### CODE REVIEW

File: pkg/loadbalancers/features/features.go [hunk 1]
```diff
@@ -24,60 +24,61 @@ import (
...
 	// TODO(shance): find a way to remove these
-	BackendService LBResource = "BackendService"
-	HealthCheck    LBResource = "HealthCheck"
-)
+	BackendService meta.Version
+	HealthCheck    meta.Version
+}
+
+// All of these fields must be filled in to allow L7ILBVersions() to work
+var l7IlbVersions = ResourceVersions{
```
Comment (bowei) [inline pkg/loadbalancers/features/features.go]: you can have a 
```
var (
  ...
)
```

File: pkg/loadbalancers/features/features.go [hunk 2]
```diff
@@ -91,15 +92,48 @@ func featuresFromIngress(ing *v1beta1.Ingress) []string {
 }
 
 // versionFromFeatures returns the meta.Version required for a list of features
-func versionFromFeatures(strings []string, resource LBResource) meta.Version {
-	fs := sets.NewString(strings...)
-	if fs.HasAny(resourceToVersionMap[resource][meta.VersionAlpha]...) {
-		return meta.VersionAlpha
+func versionsFromFeatures(features []string) ResourceVersions {
```
Comment (bowei) [inline pkg/loadbalancers/features/features.go]: It's probably easier to implement it like this:
```
func NewResourceVersions() *ResourceVersions {
  return &ResourceVersions{ /* all verisonGA */ }
}

// Merge two resource versions together, returning a new one with the combined
// versions for each resource.
func (r *ResourceVersions) Merge(other *ResourceVersions) *ResourceVersions {
  return &ResourceVersions{
    UrlMap: mergeVersions(r.UrlMap, other.UrlMap),
    ...
  }
}

func mergeVersions(a, b meta.Version) meta.Version {
  if versionOrdinal(a) < versionOrdinal(b) {
    return b
  }
  return a
}

func versionOrdinal(v meta.Version) int {
  switch v {
    case meta.VersionAlpha: return 2
    case meta.VersionBeta: return 1
  }
  return 0
}
```

Comment (spencerhance) [reply pkg/loadbalancers/features/features.go]: Done

File: pkg/loadbalancers/features/features.go (hunk 1 above)
Comment (spencerhance) [reply pkg/loadbalancers/features/features.go]: Done

File: pkg/loadbalancers/features/features.go [hunk 3]
```diff
@@ -24,63 +24,78 @@ import (
...
+)
 
-	// L7 Resources
-	UrlMap           LBResource = "UrlMap"
-	ForwardingRule   LBResource = "ForwardingRule"
-	TargetHttpProxy  LBResource = "TargetHttpProxy"
-	TargetHttpsProxy LBResource = "TargetHttpsProxy"
-	SslCertificate   LBResource = "SslCertificate"
+// ResourceVersions allows you to define all the versions required for each resource
+// for a feature.  Empty fields are considered equivalent to meta.VersionGA
```
Comment (bowei) [inline pkg/loadbalancers/features/features.go]: remove "Empty fields" etc

File: pkg/loadbalancers/features/features.go [hunk 4]
```diff
@@ -91,15 +106,38 @@ func featuresFromIngress(ing *v1beta1.Ingress) []string {
 }
 
 // versionFromFeatures returns the meta.Version required for a list of features
-func versionFromFeatures(strings []string, resource LBResource) meta.Version {
-	fs := sets.NewString(strings...)
-	if fs.HasAny(resourceToVersionMap[resource][meta.VersionAlpha]...) {
-		return meta.VersionAlpha
+func versionsFromFeatures(features []string) *ResourceVersions {
+	result := NewResourceVersions()
+	if len(features) == 0 {
```
Comment (bowei) [inline pkg/loadbalancers/features/features.go]: if is not needed i think

File: pkg/loadbalancers/features/features.go [hunk 5]
```diff
@@ -91,15 +106,38 @@ func featuresFromIngress(ing *v1beta1.Ingress) []string {
...
+		return 2
+	case meta.VersionBeta:
+		return 1
 	}
-	return meta.VersionGA
+
+	return 0
+}
+
+// Figure out the correct version
```
Comment (bowei) [inline pkg/loadbalancers/features/features.go]: // mergeVersions returns the newer API (e.g. mergeVersions(alpha, GA) = alpha).

File: pkg/loadbalancers/features/features_test.go [hunk 6]
```diff
@@ -93,71 +109,154 @@ func TestScopeFromIngress(t *testing.T) {
...
-		features   []string
-		versionMap map[LBResource]meta.Version
+		desc     string
+		features []string
+		expected *ResourceVersions
 	}{
 		{
 			desc: "No Features",
-			versionMap: map[LBResource]meta.Version{
+			expected: &ResourceVersions{
```
Comment (bowei) [inline pkg/loadbalancers/features/features_test.go]: NewResourceVersions()

File: pkg/loadbalancers/features/features_test.go [hunk 7]
```diff
@@ -93,71 +109,154 @@ func TestScopeFromIngress(t *testing.T) {
...
 				SslCertificate:   meta.VersionBeta,
 				BackendService:   meta.VersionBeta,
+				HealthCheck:      meta.VersionBeta,
 			},
 		},
 		{
 			desc:     "Differing versions",
 			features: []string{fakeGaFeature, fakeAlphaFeatureUrlMapOnly},
-			versionMap: map[LBResource]meta.Version{
+			expected: &ResourceVersions{
```
Comment (bowei) [inline pkg/loadbalancers/features/features_test.go]: You can do this neat one-liner:

`NewResourceVersions().Merge(&ResourceVersions{UrlMap:meta.VersionAlpha})`

if we change versionOrdinal to return -1 for ""

**Comment by bowei** [comment]
Just a couple of things and this one is good for merging.

File: pkg/loadbalancers/features/features_test.go (hunk 7 above)
Comment (spencerhance) [reply pkg/loadbalancers/features/features_test.go]: Done.  Also I realized how much I miss one-liners from the python world.

File: pkg/loadbalancers/features/features_test.go (hunk 6 above)
Comment (spencerhance) [reply pkg/loadbalancers/features/features_test.go]: Done

File: pkg/loadbalancers/features/features.go (hunk 5 above)
Comment (spencerhance) [reply pkg/loadbalancers/features/features.go]: Done

File: pkg/loadbalancers/features/features.go (hunk 4 above)
Comment (spencerhance) [reply pkg/loadbalancers/features/features.go]: Done

File: pkg/loadbalancers/features/features.go (hunk 3 above)
Comment (spencerhance) [reply pkg/loadbalancers/features/features.go]: Done

**Pull request merged**